### PR TITLE
Fixes #3254: fixes memory leak in ObserverRegistry

### LINKED ISSUE #3254: Memory leak in ObserverRegistry
`ObserverRegistry` stores observers in a `WeakHashMap` and does not remove values explicitly, relying on the weak references for cleanup.  

Buried deep in the `WeakHashMap` javadoc is a warning that the “value objects in a `WeakHashMap` are held by ordinary strong references. Thus care should be taken to ensure that value objects do not strongly refer to their own keys, either directly or indirectly, since that will prevent the keys from being discarded.”

Values of `ObserverRegistry` maps (`LifecycleBoundObserver` and `ViewBoundObserver`) both hold strong references to their observers, preventing deallocation.  This is leaking memory on FFTV, which can be shown by comparing a heap dump from master (which depends on A-C 54) with one that depends on [this draft PR](https://github.com/mozilla-mobile/android-components/pull/3255).  It seems likely that this affects most if not all projects that depend upon A-C.

### PR DESCRIPTION
**Pull request by Baron-Severin**
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

File: components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt [hunk 1]
```diff
@@ -22,8 +23,8 @@ import java.util.WeakHashMap
  */
 class ObserverRegistry<T> : Observable<T> {
     private val observers = mutableSetOf<T>()
-    private val lifecycleObservers = WeakHashMap<T, LifecycleBoundObserver<T>>()
-    private val viewObservers = WeakHashMap<T, ViewBoundObserver<T>>()
+    private val lifecycleObservers = WeakHashMap<T, WeakReference<LifecycleBoundObserver<T>>>()
+    private val viewObservers = WeakHashMap<T, WeakReference<ViewBoundObserver<T>>>()
```
Comment (pocmo) [inline components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: Do we really need a `WeakReference` here?

Wouldn't it be enough (and more explicit) to remove the entries in `unregister()`?

Comment (pocmo) [reply components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: And the explicit "remove" may be easier to test too :)

**Comment by Baron-Severin** [comment]
> Great find!
> 
> For this to leak it seems like you also need app code to keep a strong reference to the observer (since that is the weak key). Is that the case in your app? Are you keeping an observer around? I wonder if there are follow-up leaks (of other things) in this case.
> 
> CC @csadilek since we were debugging a leak recently too and that may have contributed to it.

Thanks!

@pocmo We're not holding any strong refs to the observer, and when I build FFTV with this modified code our memory leak disappears.  It seems that the strongly referenced entry value holds a reference to its key (`protected val observer: T`), preventing it from being collected.

I'll make your suggested change, write some tests, and open this for review!

**Comment by pocmo** [comment]
>  It seems that the strongly referenced entry value holds a reference to its key (protected val observer: T), preventing it from being collected.

Oh, yes, @csadilek mentioned that too today. With that we'll always keep the values :D

File: components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt [hunk 2]
```diff
@@ -95,6 +99,8 @@ class ObserverRegistry<T> : Observable<T> {
             }
             observers.clear()
             pausedObservers.clear()
+            lifecycleObservers.clear()
+            viewObservers.clear()
```
Comment (pocmo) [inline components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: It looks like there's a second bug hiding here: First we go through all observers and remove a lifecycle observer if it exists, but we never do that for an potentially existing view observer. I'll file a follow-up issue for that. Maybe we should play safe and just call `unregister()` for every observer here. Anyhow, different issue!

File: components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt [hunk 3]
```diff
@@ -84,8 +84,12 @@ class ObserverRegistry<T> : Observable<T> {
             pausedObservers.remove(observer)
         }
 
+        // Unregister observers
         lifecycleObservers[observer]?.remove()
         viewObservers[observer]?.remove()
+        // Remove observers from map
+        lifecycleObservers.remove(observer)
+        viewObservers.remove(observer)
```
Comment (csadilek) [inline components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: I think we'd need to move all these 4 four calls up into the synchronized block as the method claims to be thread-safe, but that was also already a problem before and can be a follow-up (currently only invoked on the main thread afaik).

Comment (csadilek) [reply components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: Follow-up: https://github.com/mozilla-mobile/android-components/issues/3277

File: components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt (hunk 2 above)
Comment (csadilek) [reply components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: Follow-up: https://github.com/mozilla-mobile/android-components/issues/3278

**Pull request merged**